crates/snp_validation/src/config.rs

```diff
@@ -0,0 +1,177 @@
+#[serde(deny_unknown_fields)]
+pub struct ConfigIntermediate {
+    /// Values for `author_key_digest` in the report body.
+    /// This is the list of public keys which have signed the signing key of the Enarx binary.
+    #[serde(default)]
+    pub enarx_signer: Vec<Digest>,
+
+    /// Values for `measurement` in the report body.
+    /// This is the hash of the Enclave environment after the Enarx binary is loaded
+    /// but before any workload is loaded, so this is a hash of the Enarx binary
+    /// in memory.
+    #[serde(default)]
+    pub enarx_hash: Vec<Digest>,
+
+    /// Values for `id_key_digest` in the report body.
+    /// This is the list of public keys which have signed the Enarx binary.
+    #[serde(default)]
+    pub id_key_digest: Vec<Digest>,
+
+    /// Prohibited values for `id_key_digest` in the report body.
+    /// This is the list of public keys which have signed the Enarx binary.
+    #[serde(default)]
+    pub id_key_digest_blacklist: Vec<Digest>,
+
+    /// Minimum value for `policy.abi_major`.`policy.abi_minor`
+    #[serde(default)]
+    pub minimum_abi: VersionReq,
+
+    #[serde(default)]
+    #[serde(deserialize_with = "from_policy_string")]
+    pub policy_flags: Option<u8>,
+
+    #[serde(default)]
+    #[serde(deserialize_with = "from_platform_string")]
+    pub platform_info_flags: Option<PlatformInfoFlags>,
+}
+
+fn from_policy_string<'de, D>(deserializer: D) -> Result<Option<u8>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde::de::Error;
+
+    let s: &str = Deserialize::deserialize(deserializer)?;
+
+    let mut flags = FlagSet::<PolicyFlags>::from(PolicyFlags::Reserved);
+
+    for flag in s.to_string().split('|') {
+        match flag.trim() {
+            "Debug" => {
+                flags |= PolicyFlags::Debug;
+            }
+            "MigrateMA" => {
+                return Err(D::Error::custom("SNP migration not supported"));
+            }
+            "SingleSocket" => {
+                flags |= PolicyFlags::SingleSocket;
+            }
+            "SMT" => {
+                flags |= PolicyFlags::SMT;
+            }
+            _ => return Err(D::Error::custom(format!("unknown policy flag '{}'", flag))),
+        }
+    }
+
+    Ok(Some(flags.bits()))
+}
+
+fn from_platform_string<'de, D>(deserializer: D) -> Result<Option<PlatformInfoFlags>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    use serde::de::Error;
+
+    let s: &str = Deserialize::deserialize(deserializer)?;
+
+    let flag = match s.trim() {
+        "SME" => PlatformInfoFlags::SME,
+        "TSME" => PlatformInfoFlags::TSME,
+        _ => {
+            return Err(D::Error::custom(format!(
+                "unknown platform info flag '{}'",
+                s
+            )))
+        }
+    };
+
+    Ok(Some(flag))
+}
+
+#[derive(Clone, Debug, Default, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    #[serde(flatten)]
+    pub config: ConfigIntermediate,
```

Comment:
This is extremely awkward typing, there should only be one `Config`. I don't have time to read through the whole implementation of this massive PR to understand why exactly this was done, but you are doing something wrong if you need this kind of crutches

Reply:
Fixed